src/lib/actions-watcher.js

```diff
@@ -75,30 +75,46 @@ async function buildAndDeploy (watcherOptions) {
 function createChangeHandler (watcherOptions) {
   const { watcher, log } = watcherOptions
 
-  let running = false
-  let changed = false
+  let deploymentInProgress = false
+  let fileChanged = false
+  let undeployedFile = ''
 
   return async (filePath) => {
-    if (running) {
+    if (deploymentInProgress) {
       aioLogger.debug(`${filePath} has changed. Deploy in progress. This change will be deployed after completion of current deployment.`)
-      changed = true
+      undeployedFile = filePath
+      fileChanged = true
       return
     }
-    running = true
+    deploymentInProgress = true
     try {
       aioLogger.debug(`${filePath} has changed. Redeploying actions.`)
+      const actionName = getActionNameFromPath(filePath)
+      watcherOptions.config.filterActions = [actionName]
       await buildAndDeploy(watcherOptions)
-      aioLogger.debug('Deployment successful.')
+      aioLogger.debug(`Deployment successful for', ${actionName}.`)
     } catch (err) {
       log('  -> Error encountered while deploying actions. Stopping auto refresh.')
       aioLogger.debug(err)
       await watcher.close()
     }
-    if (changed) {
+    if (fileChanged) {
       aioLogger.debug('Code changed during deployment. Triggering deploy again.')
-      changed = running = false
-      await createChangeHandler(watcherOptions)(filePath)
+      fileChanged = deploymentInProgress = false
+      await createChangeHandler(watcherOptions)(undeployedFile)
     }
-    running = false
+    deploymentInProgress = false
   }
 }
+
+/**
+ * Util function which extract the actionName from the filePath.
+ *
+ * @param {string} filePath  path of the file
+ * @returns {string}  name of the action
+ */
+function getActionNameFromPath (filePath) {
```

Comment:
that won't always work unfortunately, the action name does not have to be related to the path, you will have to match the path to the action name from the info in the manifest, and even then beware, the manifest can define multiple actions for the same action path (with different inputs for example)

Reply:
Hmm, having multiple actions with the same action path would be a hassle.
I suppose this approach won't work at all.

Reply:
"the manifest can define multiple actions for the same action path (with different inputs for example)"

@moritzraho Any idea where could I find such an example?

Reply:
Let's assume the following manifest actions
```
actions:
  hello:
     function: 'my/path/to/single-action.js'
     input:
       default: 'param'
  hello-2:
     function: 'my/path/to/single-action.js'
     input:
       default: 'paramvalue2'
```


If `'my/path/to/single-action.js'` changes then you will have to redeploy the two actions based on the manifest information

Reply:
In this case, I would return an array with all of the matches.